src/modm/platform/eth/stm32/eth_impl.hpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2020, Mike Wolfram
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_ETH_HPP
+#	error	"Don't include this file directly, use 'eth.hpp' instead!"
+#endif
+
+namespace modm::platform {
+
+
+template <class PHY>
+void
+Eth<PHY>::start()
+{
+	// transmission enable
+	uint32_t tmp = ETH->MACCR | ETH_MACCR_TE;
+	writeMACCR(tmp);
+
+	// reception enable
+	tmp = ETH->MACCR | ETH_MACCR_RE;
+	writeMACCR(tmp);
+
+	// flush transmission fifo
+	tmp = ETH->DMAOMR | ETH_DMAOMR_FTF;
+	writeDMAOMR(tmp);
+
+	// DMA transmission enable
+	tmp = ETH->DMAOMR | ETH_DMAOMR_ST;
+	writeDMAOMR(tmp);
+
+	// DMA reception enable
+	tmp = ETH->DMAOMR | ETH_DMAOMR_SR;
+	writeDMAOMR(tmp);
+}
+
+template <class PHY>
+void
+Eth<PHY>::stop()
+{
+	// DMA transmission disable
+	uint32_t tmp = ETH->DMAOMR & ~ETH_DMAOMR_ST;
+	writeDMAOMR(tmp);
+
+	// DMA reception disable
+	tmp = ETH->DMAOMR & ~ETH_DMAOMR_SR;
+	writeDMAOMR(tmp);
+
+	// reception disable
+	tmp = ETH->MACCR & ~ETH_MACCR_RE;
+	writeMACCR(tmp);
+
+	// flush transmission fifo
+	tmp = ETH->DMAOMR | ETH_DMAOMR_FTF;
+	writeDMAOMR(tmp);
+
+	// transmission disable
+	tmp = ETH->MACCR & ~ETH_MACCR_TE;
+	writeMACCR(tmp);
+}
+
+template <class PHY>
+void
+Eth<PHY>::configureMacAddresses()
+{
+	static constexpr uint32_t ETH_MAC_ADDR_HBASE { ETH_MAC_BASE + 0x40 };  /* Ethernet MAC address high offset */
+	static constexpr uint32_t ETH_MAC_ADDR_LBASE { ETH_MAC_BASE + 0x44 };  /* Ethernet MAC address low offset */
+
+	static constexpr MacAddress zeroMac { 0 };
+
+	for (std::size_t i = 0; i < macAddresses.size(); ++i) {
+		auto const &macAddress = macAddresses[i];
+		if (std::memcmp(zeroMac.data(), macAddress.data(), macAddress.size()) == 0)
```

Comment:
This could be `==` instead of `memcmp` since we have a `std::array`. Code-gen will also be `memcmp` in that case.